### PR TITLE
docs(specs): add Toast spec

### PR DESCRIPTION
**Pull request by stonebk**
Tip: Switch to rich diff for easier reading:
Screen Shot 2019-07-10 at 11 17 30 AM

### CODE REVIEW

**Comment by WattyRev** [comment]
I think Toasts are a bit special in that they should be able to be invoked programmatically without the context of a component. For example, maybe I have a file that provides methods for communicating with some specific backend API. I may want that file to be able to throw up a toast if an API request fails.

With the current spec, it looks like you need to wrap a component with the HOC in order to gain access to the `enqueueToast` method, so I would have to explicitly handle these api failures in every component that makes the api calls, making it harder to have centralized error handling on an API basis. Any thoughts on how to provide easy access to `enqueueToast` outside of the general React Component flow?

File: specs/Toast.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+#### `<ToastProvider>`
+
+There can be multiple applications on the page, but all apps should share the same notification area. The first app to initialize should set up the portal and configuration; all subsequent providers will use the portal and configuration of the first to initialize.
+
+| Name | Type | Default | Description |
+| ---- | ---- | ------- | ----------- |
+| `placement` | `string` | | Where the region will exist with respect to the viewport. e.g. `"top-start"`, `"top"`, `"top-end"`, `"bottom-start"`, `"bottom"`, `"bottom-end"` |
+| `maxOpen` | `number` | 3 | The maximum number of toasts that can be open at once. |
+| `autoClose` | `number` | 5000 | The delay in milliseconds before automatically closing a toast. If set to 0, the toast must be manually dismissed. |
```
Comment (stonebk) [inline specs/Toast.md]: This probably needs to be customizable at the Toast level as some Toasts might need to linger longer, or not close at all.

**Comment by stonebk** [comment]
> I think Toasts are a bit special in that they should be able to be invoked programmatically without the context of a component. For example, maybe I have a file that provides methods for communicating with some specific backend API. I may want that file to be able to throw up a toast if an API request fails.
> 
> With the current spec, it looks like you need to wrap a component with the HOC in order to gain access to the `enqueueToast` method, so I would have to explicitly handle these api failures in every component that makes the api calls, making it harder to have centralized error handling on an API basis. Any thoughts on how to provide easy access to `enqueueToast` outside of the general React Component flow?

I think toast messaging is still part of the view which should come from a component somewhere in the react application. I don't think API logic shouldn't be altering the view directly, but updating state somewhere (redux store, local component state) that triggers a component update. If you have a lot of components making API requests, I would definitely try to centralize that logic, and then you can create a centralized component view around that state.

I honestly don't really know what it would look like to provide easy access outside of React component flow -- IMHO, that feels a bit like a workaround for poor application architecture. Are there any patterns you have seen that provide something like this? I'm definitely open to suggestions.

**Comment by WattyRev** [comment]
> > I think Toasts are a bit special in that they should be able to be invoked programmatically without the context of a component. For example, maybe I have a file that provides methods for communicating with some specific backend API. I may want that file to be able to throw up a toast if an API request fails.
> > With the current spec, it looks like you need to wrap a component with the HOC in order to gain access to the `enqueueToast` method, so I would have to explicitly handle these api failures in every component that makes the api calls, making it harder to have centralized error handling on an API basis. Any thoughts on how to provide easy access to `enqueueToast` outside of the general React Component flow?
> 
> I think toast messaging is still part of the view which should come from a component somewhere in the react application. I don't think API logic shouldn't be altering the view directly, but updating state somewhere (redux store, local component state) that triggers a component update. If you have a lot of components making API requests, I would definitely try to centralize that logic, and then you can create a centralized component view around that state.
> 
> I honestly don't really know what it would look like to provide easy access outside of React component flow -- IMHO, that feels a bit like a workaround for poor application architecture. Are there any patterns you have seen that provide something like this? I'm definitely open to suggestions.

We are using React Toastify on my team: https://www.npmjs.com/package/react-toastify

The toast itself is still in the component architecture as it should be, but you can trigger a toast to from anywhere by importing and calling the `toast` method. I think this makes sense in my team's use case because we display technical errors to our users, so it makes it easier to triage/debug issues. When we have an error that should be displayed to the user, we can do so right where we expect the error to happen rather than waiting for it to get down to some part of the UI that is capable of triggering it.

**Comment by stonebk** [comment]
> We are using React Toastify on my team: https://www.npmjs.com/package/react-toastify
> 
> The toast itself is still in the component architecture as it should be, but you can trigger a toast to from anywhere by importing and calling the `toast` method. I think this makes sense in my team's use case because we display technical errors to our users, so it makes it easier to triage/debug issues. When we have an error that should be displayed to the user, we can do so right where we expect the error to happen rather than waiting for it to get down to some part of the UI that is capable of triggering it.

That definitely looks a lot easier to use. Thanks, I'll play around with this a bit!

**Comment by stonebk** [comment]
I’ve moved forward with the toast provider functionality as I have it specced out. My thought is that this is a clean implementation that does not rely on any global state. I think for the global use-case, you could create a small wrapper around this implementation.

**Pull request merged**